controllers/host_maintenance_controller.go

```diff
@@ -71,11 +196,215 @@ func (r *HMCMaintenanceReconciler) Reconcile(ctx context.Context, request ctrl.R
+		For(&infrav1beta1.MaasMachine{}).
+		WithEventFilter(predicate.Funcs{
+			CreateFunc: func(e event.CreateEvent) bool {
+				return r.isHostMachine(e.Object)
+			},
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				return r.isHostMachine(e.ObjectNew)
+			},
+			DeleteFunc: func(e event.DeleteEvent) bool {
+				return r.isHostMachine(e.Object)
+			},
+			GenericFunc: func(e event.GenericEvent) bool {
+				return r.isHostMachine(e.Object)
+			},
+		}).
 		For(&corev1.ConfigMap{}).
 		WithOptions(controller.Options{MaxConcurrentReconciles: 1}).
 		Complete(r)
 }
+
+// isHostMachine checks if the MaasMachine is a host (not a VM)
+func (r *HMCMaintenanceReconciler) isHostMachine(obj client.Object) bool {
+	maasMachine, ok := obj.(*infrav1beta1.MaasMachine)
+	if !ok {
+		return false
+	}
+
+	// Check if this is a host machine (not a VM)
+	// VMs typically have a parent reference, hosts don't
+	return maasMachine.Spec.Parent == nil || *maasMachine.Spec.Parent == ""
+}
+
+// checkEvacuationGates checks if evacuation criteria are met
+func (r *HMCMaintenanceReconciler) checkEvacuationGates(ctx context.Context, maasMachine *infrav1beta1.MaasMachine, log logr.Logger) (bool, error) {
+	if maasMachine.Spec.SystemID == nil {
+		return false, fmt.Errorf("maasMachine has no systemID")
+	}
+	hostSystemID := *maasMachine.Spec.SystemID
+
+	// Gate 1: Check if host is empty (no VMs running)
+	hostEmpty, err := r.isHostEmpty(ctx, hostSystemID, log)
+	if err != nil {
+		return false, fmt.Errorf("failed to check if host is empty: %w", err)
+	}
+
+	if !hostEmpty {
+		log.Info("Host not empty, evacuation blocked", "host", hostSystemID)
+		return false, nil
+	}
+
+	// Gate 2: Check per-WLC ready-op-<uuid> tags on CP VMs
+	wlcReady, err := r.checkWLCReadyTags(ctx, hostSystemID, log)
+	if err != nil {
+		return false, fmt.Errorf("failed to check WLC ready tags: %w", err)
+	}
+
+	if !wlcReady {
+		log.Info("WLC ready tags not met, evacuation blocked", "host", hostSystemID)
+		return false, nil
+	}
+
+	log.Info("All evacuation gates met", "host", hostSystemID)
+	return true, nil
+}
+
+// isHostEmpty checks if the host has no running VMs
+func (r *HMCMaintenanceReconciler) isHostEmpty(ctx context.Context, hostSystemID string, log logr.Logger) (bool, error) {
+	vms, err := r.inventoryService.ListHostVMs(hostSystemID)
+	if err != nil {
+		return false, fmt.Errorf("failed to list host VMs: %w", err)
+	}
+
+	// Check if any VMs are in running state
+	for _, vm := range vms {
+		if vm.PowerState == "on" || vm.PowerState == "running" {
+			log.Info("Host has running VM", "host", hostSystemID, "vm", vm.SystemID, "powerState", vm.PowerState)
+			return false, nil
+		}
+	}
+
+	log.Info("Host is empty", "host", hostSystemID, "vmCount", len(vms))
+	return true, nil
+}
+
+// checkWLCReadyTags checks if all WLC clusters have ready-op-<uuid> tags
+func (r *HMCMaintenanceReconciler) checkWLCReadyTags(ctx context.Context, hostSystemID string, log logr.Logger) (bool, error) {
```

Comment:
Check ready tags for VM.

- Create a backup of all WLC VMs+Clusters
- Check all registred hosts in MAAS lab
- Check parent machine of hosts, for tags
- Filter ClusterTag+CPTag+OpID tag, cross check with tags current hosts had
- If tags maches with, that identified VM is moved to new host